googleapis/src/main/java/io/grpc/googleapis/GoogleCloudToProdNameResolver.java

```diff
@@ -156,29 +156,35 @@ private void resolve() {
     class Resolve implements Runnable {
       @Override
       public void run() {
-        String zone;
-        boolean supportIpv6;
         ImmutableMap<String, ?> rawBootstrap = null;
-        try {
-          zone = queryZoneMetadata(METADATA_URL_ZONE);
-          supportIpv6 = queryIpv6SupportMetadata(METADATA_URL_SUPPORT_IPV6);
-          rawBootstrap = generateBootstrap(zone, supportIpv6);
-        } catch (IOException e) {
-          listener.onError(Status.INTERNAL.withDescription("Unable to get metadata").withCause(e));
-        } finally {
-          final ImmutableMap<String, ?> finalRawBootstrap = rawBootstrap;
-          syncContext.execute(new Runnable() {
-            @Override
-            public void run() {
-              if (!shutdown && finalRawBootstrap != null) {
+        // User provided bootstrap configs are only supported with federation. If federation is
+        // not enabled or there is no user provided config, we set a custom bootstrap override.
+        // Otherwise, we don't set the override, which will allow a user provided bootstrap config
+        // to take effect.
+        if (!enableFederation || !xdsBootstrapProvided) {
+          try {
+            rawBootstrap = generateBootstrap(queryZoneMetadata(METADATA_URL_ZONE),
+                queryIpv6SupportMetadata(METADATA_URL_SUPPORT_IPV6));
+          } catch (IOException e) {
+            listener.onError(
+                Status.INTERNAL.withDescription("Unable to get metadata").withCause(e));
+          }
+        }
+
+        final ImmutableMap<String, ?> finalRawBootstrap = rawBootstrap;
```

Comment:
why remove finally block? 
otherwise it looks fine.

Reply:
Good catch, I simplified too much, that could leak a runtime exception. Added the finally block back.